pkg/controller/deployment_inplace.go

```diff
@@ -0,0 +1,559 @@
+					return addedNewReplicasCount > 0, err
+				}
+
+				return scaled, err
+			}
+
+			// uncordon the node since the ownership of the machine has been transferred to the new machien set.
+			node.Spec.Unschedulable = false
+			_, err = dc.targetCoreClient.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+			if err != nil {
+				return false, fmt.Errorf("failed to uncordon the node %s: %w", node.Name, err)
+			}
+
+			transferredMachineCount++ // scale down the old machine set.
+			addedNewReplicasCount++   // scale up the new machine set.
+		}
+
+		klog.V(3).Infof("scale down machine set %s, transffered replicas to new machine set %d", is.Name, transferredMachineCount)
+		_, _, err = dc.scaleMachineSetAndRecordEvent(ctx, is, is.Spec.Replicas-transferredMachineCount, deployment)
+		if err != nil {
+			klog.V(3).Infof("scale down failed %s", err)
+			return addedNewReplicasCount > 0, err
+		}
+	}
+
+	klog.V(3).Infof("scale up the new machine set %s by %d", newIS.Name, addedNewReplicasCount)
+	scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newIS, newIS.Spec.Replicas+addedNewReplicasCount, deployment)
+	return scaled, err
+}
+
+func (dc *controller) reconcileOldMachineSetsInPlace(ctx context.Context, allISs []*v1alpha1.MachineSet, oldISs []*v1alpha1.MachineSet, newIS *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
+	oldMachinesCount := GetReplicaCountForMachineSets(oldISs)
+	if oldMachinesCount == 0 {
+		// Can't scale down further
+		return false, nil
+	}
+
+	// If maxSurge is defined, there will be machines left in the old machine set that do not require an update
+	// because the new machine set already has the required replicas due to the additional machines added as per maxSurge.
+	// In that case we simply scale down the old machine set to zero.
+	if (newIS.Spec.Replicas) == (deployment.Spec.Replicas) {
+		// Scale down old machine sets to zero.
+		for _, is := range oldISs {
+			_, _, err := dc.scaleMachineSetAndRecordEvent(ctx, is, 0, deployment)
+			if err != nil {
+				return false, err
+			}
+		}
+		return true, nil
+	}
+
+	allMachinesCount := GetReplicaCountForMachineSets(allISs)
+	klog.V(3).Infof("New machine set %s has %d available machines.", newIS.Name, newIS.Status.AvailableReplicas)
+	maxUnavailable := MaxUnavailable(*deployment)
+
+	minAvailable := (deployment.Spec.Replicas) - maxUnavailable
+	newISUnavailableMachineCount := (newIS.Spec.Replicas) - newIS.Status.AvailableReplicas
+	oldISsMachinesUndergoingUpdate, err := dc.getMachinesUndergoingUpdate(oldISs)
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(3).Infof("allMachinesCount:%d,  minAvailable:%d,  newISUnavailableMachineCount:%d,  oldISsMachineInUpdateProcess:%d", allMachinesCount, minAvailable, newISUnavailableMachineCount, oldISsMachinesUndergoingUpdate)
+
+	maxUpdatePossible := allMachinesCount - minAvailable - newISUnavailableMachineCount - oldISsMachinesUndergoingUpdate
```

Comment:
Please move this computation to `selectMachineForUpdate()` as that is the most semantic place for this computation